---
title: "QTL2 Mediation"
author: "Brian Yandell"
date: "7/5/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(ggplot2)
```

### Specific analysis setup

These set up the phenotype and its peak location for further analysis.

```{r}
chr_id <- "11"
scan_window <- c(96,98)
pheno_name <- "VC_top_distance_first4"
peak_Mbp <- mean(scan_window)
window_Mbp <- diff(scan_window) / 2
```

## Setup

```{r}
local <- TRUE # consider local QTL (from mRNA traits)
other <- FALSE # consider other types of phenotypes if TRUE
qtls <- 1 # or 2 for number of drivers
```

The `qtl2shiny` package uses a data frame of project information (folder locations).

```{r}
dirpath <- "../../qtl2shinyApp"
project_info <- data.frame(taxa = "CCmouse",
                           project = "Recla",
                           directory = "qtl2shinyData")
```

These encode a hierarchy:

```{r}
datapath <- file.path(file.path(dirpath, project_info$directory))
taxapath <- file.path(datapath, project_info$taxa)
projectpath <- file.path(taxapath, project_info$project)
```

Information for all phenotypes: all `covar`iates and the physical map `pmap`.

```{r}
covar <- readRDS(file.path(projectpath, "covar.rds"))
str(covar)
```

```{r}
pmap <- readRDS(file.path(projectpath, "pmap.rds"))
sapply(pmap, range)
```

Now subset on region of interest:

```{r}
pmap <- pmap[[chr_id]]
```

These are used to find objects that are not immediately loaded, including 
genotype probabilities and gene and variant information. See next section on query routines.

The phenotype data are extracted here.

```{r}
pheno_data <- readRDS(file.path(projectpath, "pheno_data.rds"))
dim(pheno_data)
colnames(pheno_data)
```

The `peaks` and `analyses_tbl` have information on phenotypes. The `peaks` has location of all previously found peaks, used to find what maps in a region. The `analyses_tbl` has details of transformations and covariates.
Phenotypes are ordered into groups and types, which is useful for searching and for determining
sets of covariates and transformations when there are many phenotypes to consider.

```{r}
peaks <- readRDS(file.path(projectpath, "peaks.rds"))
str(peaks)
```

```{r}
analyses_tbl <- readRDS(file.path(projectpath, "analyses.rds"))
str(analyses_tbl)
```

### Query routines

The query routines embed the location and database information for genes, variants, probabilities and mRNA data, which are generally very large and only needed for a subset. All have first three arguments
`(chr, start, end)`

They are used in `read_project` and `read_query_rds`, but we make their use explicit here
for encoding locations in query functions.

```{r}
query_genes <- qtl2::create_gene_query_func(file.path(taxapath, "mouse_genes.sqlite"))
query_variants <- qtl2::create_variant_query_func(file.path(taxapath, "cc_variants.sqlite"))
```

```{r}
args("query_genes")
```

```{r}
str(query_genes(chr_id, scan_window[1], scan_window[2]))
```

```{r}
args("query_variants")
```

```{r}
str(query_variants(chr_id, scan_window[1], scan_window[2]))
```

The `query_probs` assumes a `probdir` (default `genoprob`), which contains the FST `method` files `fst_probs.rds` (for genotype probabilities) and `fst_aprobs.rds` (for allele probabilities). These are FST master files, which point to individual FST databases by chromosome, such as `aprobs_1.fst` and `probs_1.fst`.

```{r}
query_probs <- qtl2pattern::create_probs_query_func(projectpath)
```

```{r}
args("query_probs")
```

```{r}
get("method_val", env = environment(query_probs))
```

```{r}
get("probdir_val", env = environment(query_probs))
```

```{r}
probs_obj <- query_probs(chr_id, scan_window[1], scan_window[2], allele = TRUE)
dim(probs_obj$probs[[1]])
```
```{r}
range(probs_obj$map)
```

```{r}
dim(query_probs(chr_id, scan_window[1], scan_window[2], allele = FALSE)$probs[[1]])
```

The `query_mrna` assumes a `mrnadir` (default `RNAseq`), which should contain the following files:

```
annot.mrna.rds
annot.samples.rds
expr.mrna.fst
peaks.mrna.fst
```

```{r}
query_mrna <- qtl2shiny::create_mrna_query_func(projectpath)
```

The function has additional arguments `local` (use only mRNA whose genes are in the region if `TRUE`)
and `qtl` (use only mRNA whose QTL peaks are in the region if `TRUE`).

```{r}
args("query_mrna")
```

```{r}
get("mrnadir_val", env = environment(query_mrna))
```

There are no mRNA data for the Recla dataset, so this will not be used below.
The following will create an error for these data.

```
str(query_mrna(chr_id, scan_window[1], scan_window[2]))
```


## Phenotype Mediation

Phenotype mediation uses as mediators other phenotypes in the region defined by `chr_id` and `scan_window`. This is all bundled in the routine `comediator_region`, using `project_info` and `analyses_tbl` settings. It then calls `pheno_region` with `peaks` and `qtls`.

```
comed_ls <- 
    comediator_region(pheno_name, chr_id, scan_window, 
                      covar, analyses_tbl, peaks, 
                      qtls, pmap, project_info)
```

```{r}
# This is specific to CCmouse.
peaks <- dplyr::filter(peaks,
                       .data$pheno != pheno_name,
                       !(.data$pheno_group == "Islet.mRNA"))
  
# Filter peaks and analyses to region and drop pheno_name
peaks_local <- dplyr::filter(peaks,
                             .data$chr == chr_id,
                             .data$pos >= scan_window[1],
                             .data$pos <= scan_window[2])
analyses_df <- dplyr::filter(analyses_tbl,
                             .data$pheno %in% peaks_local$pheno,
                             .data$pheno != pheno_name)
  
# Read the phenos we need.
phenos <- analyses_df$pheno
pheno_other <- pheno_data[, phenos, drop = FALSE]

# Create comediator object.
comed_ls <- qtl2shiny:::pheno_region(
    chr_id, scan_window[1], scan_window[2], covar, pmap,
    peaks, analyses_tbl, pheno_other, drivers = qtls)
```

```{r}
med_ls <- qtl2shiny:::comediator_type(comed_ls, peaks, pheno_name, other)
```

## Expression Mediation

Need to either have `query_mrna` set up ahead of time (in 'RDS' format) or create here.

```
expr_ls <- get_expr_region(chr_id, scan_window[1], scan_window[2], covar, pmap, 
                           drivers = qtls,
                           query_mrna = query_mrna)
```

## Mediation Test

Mediation test is performed in shinyMediate. Input includes `med_ls` (from other phenotypes) or `expr_ls` (from mRNA expression data).

```{r}
kinship <- readRDS(file.path(projectpath, "kinship.rds"))[[chr_id]]
```

The position for target mediation is by default the center of the map in this region.
However, it might be improved by looking at peak for `pheno_name`.

```{r}
pos_Mbp <- mean(range(probs_obj$map))
```

```{r}
peak_mar <- qtl2::find_marker(probs_obj$map, chr_id, pos_Mbp)
geno_max <- subset(probs_obj$probs, chr = chr_id, mar = peak_mar)[[1]][,,1]
```

```{r}
cov_tar <- qtl2shiny:::get_covar(
  covar, 
  dplyr::filter(analyses_tbl, pheno == pheno_name))
driver_med <- probs_obj$probs[[chr_id]]
```

```{r}
phe_mx <- pheno_data[, pheno_name, drop = FALSE]
```

```{r}
med_scan <- intermediate::mediation_scan(
  target = phe_mx,
  mediator = med_ls[[1]],
  driver = geno_max,
  annotation = med_ls[[2]],
  covar = cov_tar,
  kinship = kinship,
  test = "wilc",
  fitFunction =  qtl2shiny::fitQtl2)
```

```{r}
summary(med_scan, minimal = TRUE)
```

```{r}
ggplot2::autoplot(med_scan, size = 3)
```

```{r}
med_test <- intermediate::mediation_test(
  target = phe_mx,
  mediator = med_ls[[1]],
  driver = geno_max,
  annotation = med_ls[[2]],
  covar_tar = cov_tar,
  covar_med = med_ls$covar,
  kinship = kinship,
  driver_med = driver_med,
  test = "wilc",
  pos = pos_Mbp,
  data_type = "phenotype",
  fitFunction =  qtl2shiny::fitQtl2)
```

```{r}
summary(med_test)
```

```{r}
ggplot2::autoplot(med_test, size = 3, alpha = 1)
```

## Mediation Triad

Called from shinyTriad

Arguments passed to shinyTriad, shinySNPSetup, shinySNPProbs.
Need to check if using probs or aprobs for snpprobs.

```{r}
snpinfo <- query_variants(chr_id, scan_window[1], scan_window[2])
snpprobs <- qtl2shiny:::get_snpprobs(chr_id, peak_Mbp, window_Mbp,
                   pheno_name, 
                   probs_obj$probs,
                   probs_obj$map,
                   snpinfo)
```

```{r}
snp_scan_obj <- qtl2shiny:::scan1_covar(pheno_data, cov_tar,
                                        snpprobs$snpprobs, kinship,
                                        analyses_tbl,
                                        sex_type = "all")
```

```{r}
ggplot2::autoplot(snp_scan_obj, snpprobs$snpinfo, patterns = "hilit", drop_hilit = 3,
                  haplos = snpprobs$haplos)
```

```{r}
top_snps_tbl <- qtl2pattern::top_snps_pattern(
  snp_scan_obj,
  snpprobs$snpinfo)
```

```{r}
summary(top_snps_tbl)
```


```{r}
patterns <- summary(top_snps_tbl)
haplos <- LETTERS[1:8]

triad <- unique(med_test$best$triad)

# Pick top trait with causal call.
med_name <- dplyr::filter(med_test$best, .data$triad == "causal")[["id"]][1]
pattern <- dplyr::filter(patterns, .data$pheno == med_name)$pattern[1]

sdps <- unique(dplyr::filter(patterns, .data$pheno == med_name)$sdp)

sdp <- sdps[qtl2pattern::sdp_to_pattern(sdps, haplos) == pattern]
id <- med_ls[[2]]$id[med_ls[[2]][["id"]] == med_name]

cov_tar <- qtl2shiny:::covar_df_mx(cov_tar)
cov_med <- qtl2shiny:::covar_df_mx(med_ls$covar)
```

```{r}
med_triad <- intermediate::mediation_triad(target = phe_mx,
                              mediator = med_ls[[1]][, id, drop = FALSE],
                              driver = geno_max, 
                              covar_tar = cov_tar,
                              covar_med = cov_med,
                              kinship = kinship,
                              sdp = sdp, allele = TRUE,
                              fitFunction = qtl2shiny::fitQtl2)
```

```{r}
summary(med_triad)
```

```{r}
intermediate::ggplot_mediation_triad(med_triad)
```

